Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.cpp

```diff
@@ -36,35 +42,48 @@ namespace CSSPropertyParserHelpers {
 // MARK: <url>
 // https://drafts.csswg.org/css-values/#urls
 
-StringView consumeURLRaw(CSSParserTokenRange& range)
+std::optional<CSS::URL> consumeURLRaw(CSSParserTokenRange& range, CSS::PropertyParserState& state)
 {
     auto& token = range.peek();
     if (token.type() == UrlToken) {
+        auto result = CSS::completeURL(token.value().toString(), state.context);
+        if (!result)
+            return { };
         range.consumeIncludingWhitespace();
-        return token.value();
+        return result;
     }
 
-    if (token.functionId() == CSSValueUrl) {
-        auto rangeCopy = range;
-        auto args = rangeCopy.consumeBlock();
-        auto& next = args.consumeIncludingWhitespace();
-        if (next.type() == BadStringToken || !args.atEnd())
-            return StringView();
-        ASSERT(next.type() == StringToken);
-        range = rangeCopy;
-        range.consumeWhitespace();
-        return next.value();
+    switch (token.functionId()) {
+    case CSSValueUrl: {
+        CSSParserTokenRangeGuard guard { range };
+
+        auto args = consumeFunction(range);
+
+        auto string = consumeStringRaw(args);
+        if (string.isNull())
+            return { };
+        auto result = CSS::completeURL(string.toString(), state.context);
+        if (!result || !args.atEnd())
+            return { };
+
+        guard.commit();
+
+        return result;
+    }
+
+    default:
+        break;
     }
 
     return { };
 }
 
-RefPtr<CSSPrimitiveValue> consumeURL(CSSParserTokenRange& range)
+RefPtr<CSSValue> consumeURL(CSSParserTokenRange& range, CSS::PropertyParserState& state)
 {
-    auto url = consumeURLRaw(range);
-    if (url.isNull())
+    auto url = consumeURLRaw(range, state);
```

Comment:
Another example (I didn’t comment on all of them) where I would wriggle to avoid the name `url`:
```suggestion
    auto rawURL = consumeURLRaw(range, state);
```